File: roboplan_ros_py/test/test_kinematics.py
import pytest
import numpy as np
from pathlib import Path

from roboplan import (
    get_package_share_dir,
    Scene,
    SimpleIkOptions,
)

from roboplan_ros_py.kinematics import RoboPlanIK
from roboplan_ros_py.type_conversions import se3_to_pose


@pytest.fixture
def test_scene() -> Scene:
    """Create a test scene with UR5 robot."""
    roboplan_examples_dir = Path(get_package_share_dir())
    urdf_path = roboplan_examples_dir / "ur_robot_model" / "ur5_gripper.urdf"
    srdf_path = roboplan_examples_dir / "ur_robot_model" / "ur5_gripper.srdf"
    package_paths = [roboplan_examples_dir]

    return Scene("test_scene", urdf_path, srdf_path, package_paths)


@pytest.fixture
def ik_solver(test_scene: Scene) -> RoboPlanIK:
    """Create an IK solver instance."""
    options = SimpleIkOptions()
    options.group_name = "arm"
    options.max_iters = 100
    options.step_size = 0.25
    options.damping = 0.001
    options.max_error_norm = 0.001

    return RoboPlanIK(
        scene=test_scene,
        group_name="arm",
        base_frame="base_link",
        tip_frame="tool0",
        options=options,
    )


def test_ik_solver_initialization(ik_solver: RoboPlanIK) -> None:
    """Test that IK solver initializes correctly."""
    assert ik_solver.group_name == "arm"
    assert ik_solver.base_frame == "base_link"
    assert ik_solver.tip_frame == "tool0"
    assert len(ik_solver.q_indices_) == 6
    assert ik_solver.ik_solver_ is not None


def test_solve_ik(ik_solver: RoboPlanIK, test_scene: Scene) -> None:
    # Get a collision free position and find the FK for it
    test_scene.setRngSeed(1234)
    q_full = test_scene.randomCollisionFreePositions()
    test_scene.setJointPositions(q_full)
    fk_transform = test_scene.forwardKinematics(q_full, "tool0")

    # Convert to ROS Pose and solve
    target_pose = se3_to_pose(fk_transform)
    q_indices = ik_solver.q_indices_

    # Warm start from the next random position
    seed_state = test_scene.randomCollisionFreePositions()[q_indices]
    solution = ik_solver.solve_ik(target_pose, seed_state)

    # Should find a solution that is reasonably close the the start pose
    assert solution is not None
    assert len(solution) == len(q_indices)

    solution_fk = test_scene.forwardKinematics(solution, "tool0")
    assert np.linalg.norm(solution_fk[:3, 3] - fk_transform[:3, 3]) < 0.001

File: roboplan_ros_cpp/test/test_kinematics.cpp
#include <gtest/gtest.h>
#include <roboplan_example_models/resources.hpp>
#include <roboplan_ros_cpp/kinematics.hpp>

namespace roboplan_ros_cpp {

class RoboPlanIKTest : public ::testing::Test {
protected:
  void SetUp() override {
    const auto share_prefix = roboplan::example_models::get_install_prefix() / "share";
    const auto model_prefix = share_prefix / "roboplan_example_models" / "models";
    const auto urdf_path = model_prefix / "ur_robot_model" / "ur5_gripper.urdf";
    const auto srdf_path = model_prefix / "ur_robot_model" / "ur5_gripper.srdf";
    const std::vector<std::filesystem::path> package_paths = {share_prefix};
    test_scene_ =
        std::make_shared<roboplan::Scene>("test_scene", urdf_path, srdf_path, package_paths);

    // Create an IK solver instance
    roboplan::SimpleIkOptions options;
    options.group_name = "arm";
    options.max_iters = 100;
    options.step_size = 0.25;
    options.damping = 0.001;
    options.max_error_norm = 0.001;
    ik_solver_ = std::make_shared<RoboPlanIK>(test_scene_, "arm", "base_link", "tool0", options);
  }

  void TearDown() override {}

  std::shared_ptr<roboplan::Scene> test_scene_;
  std::shared_ptr<RoboPlanIK> ik_solver_;
};

TEST_F(RoboPlanIKTest, SolveIK) {
  // Get a collision free position and find the FK for it
  test_scene_->setRngSeed(1234);
  Eigen::VectorXd q_full = test_scene_->randomCollisionFreePositions().value();
  test_scene_->setJointPositions(q_full);
  Eigen::Matrix4d fk_transform = test_scene_->forwardKinematics(q_full, "tool0");

  // Convert to ROS Pose and solve
  geometry_msgs::msg::Pose target_pose = se3ToPose(fk_transform);
  auto solution = ik_solver_->solveIK(target_pose);

  // Should find a solution that is reasonably close to the start pose
  ASSERT_TRUE(solution.has_value());
  Eigen::Matrix4d solution_fk = test_scene_->forwardKinematics(solution.value(), "tool0");
  Eigen::Vector3d position_error = solution_fk.block<3, 1>(0, 3) - fk_transform.block<3, 1>(0, 3);
  ASSERT_LT(position_error.norm(), 1e-5);
}

}  // namespace roboplan_ros_cpp
